source/program/remote_api.cpp

```diff
@@ -54,36 +76,83 @@ void PrepareThread() {
 
     int rval = nn::socket::Bind(g_TcpSocket, (struct sockaddr*)&serverAddr, sizeof(serverAddr));
     if (rval < 0) EXL_ABORT(69);
-    
+
     rval = nn::socket::Listen(g_TcpSocket, 1);
     if (rval < 0) EXL_ABORT(69);
 }
 
-void SocketSpawn(void*) {
+void AddPacketToSendBuffer(PacketBuffer buffer) {
+    nn::os::LockMutex(&sendBufferLock);
+    sendBuffer.push_back(buffer);
+    nn::os::UnlockMutex(&sendBufferLock);
+}
+
+void ReceiveLogic() {
+    // don't recv new packets while we waiting for game loop otherwise we need to make a copy of our receive buffer
+    if (readyForGameThread.load()) return;
+    memset(RecvBuffer.data(), 0, 4096);
+
+    ssize_t length = nn::socket::Recv(clientSocket, RecvBuffer.data(), 1, MSG_DONTWAIT);
+    RecvBufferLength = length;
+    if (length > 0) {
+        // data received
+        RemoteApi::ParseClientPacket();
+    }
+}
+
+void SendLogic() {
+    nn::os::LockMutex(&sendBufferLock);
+    for (PacketBuffer pb : sendBuffer) {
+        ssize_t ret = nn::socket::Send(clientSocket, pb->data(), pb->size(), 0);
+        // mark them for removal
+        if (ret > 0) {
+            pb->clear();
+        }
+    }
+    // erase everything with NULL pointer, other stays in the queue for a retry
+    sendBuffer.erase(std::remove_if(sendBuffer.begin(), sendBuffer.end(), [](PacketBuffer pb) { 
+        if (pb->size() == 0) {
```

Comment:
With `PacketBuffer` being unique_ptr, it just needs to remove the element from the vector, no free/delete needed.